NeoML/src/TraditionalML/Svm.cpp

```diff
@@ -49,71 +145,32 @@ CPtr<IModel> CSvm::Train( const IProblem& problem )
 		return nullptr;
 	}
 
-	CSvmKernel kernel( params.KernelType, params.Degree, params.Gamma, params.Coeff0 );
+	const CSvmKernel kernel( params.KernelType, params.Degree, params.Gamma, params.Coeff0 );
 
-	CSMOptimizer optimizer( kernel, problem, params.MaxIterations, params.ErrorWeight, params.Tolerance,
-		params.DoShrinking );
+	CSMOptimizer optimizer( kernel, problem, params.MaxIterations,
+		params.ErrorWeight, params.Tolerance, params.DoShrinking );
 	if( log != nullptr ) {
 		optimizer.SetLog( log );
 	}
 
-	CArray<double> alpha;
-	float freeTerm;
+	CArray<double> alpha{};
+	float freeTerm{};
 	optimizer.Optimize( alpha, freeTerm );
 
 	if( kernel.KernelType() == CSvmKernel::KT_Linear ) {
-		const int vectorCount = problem.GetVectorCount();
-		const int curThreadCount = IsOmpRelevant( vectorCount ) ? params.ThreadCount : 1;
-		const CFloatMatrixDesc matrix = problem.GetMatrix();
-		NeoAssert( matrix.Height == problem.GetVectorCount() );
-		NeoAssert( matrix.Width == problem.GetFeatureCount() );
-
-		CArray<CFloatVector> planeReduction;
-		planeReduction.Add( CFloatVector( problem.GetFeatureCount() + 1, 0.f ), curThreadCount );
-
-		NEOML_OMP_NUM_THREADS( curThreadCount )
-		{
-			const int threadNumber = OmpGetThreadNum();
-			CFloatVector& plane = planeReduction[threadNumber];
-
-			int index = 0;
-			int count = 0;
-			if( OmpGetTaskIndexAndCount( vectorCount, index, count ) ) {
-				for( int i = 0; i < count; i++ ) {
-					float alphaValue = static_cast<float>( alpha[index] * problem.GetBinaryClass(index) );
-					CFloatVectorDesc desc;
-					matrix.GetRow( index, desc );
-					plane.MultiplyAndAdd( desc, alphaValue );
-					index++;
-				}
-			}
-		}
+		CFindPlanesThreadTask planeTask( params.ThreadCount, problem, alpha );
+		NEOML_NUM_THREADS( *threadPool, &planeTask, []( int threadIndex, void* ptr ) {
```

Comment:
You don't check for the case when the task is too small for multithreading (see `curThreadCount` in original).

You can make a lambda (or separate function)
```
/*lambda-or-function*/( IThreadPool* pool, IThreadTask* task )
{
    if( task->Size() < multithreadThreshold ) {
        task->Run( 0, 0, task->Size() );
    } else {
        NEOML_NUM_THREADS( *threadPool, &distTask, []( int threadIndex, void* ptr ) {
            plane += planeReduction[i];
            ( ( IThreadTask* )ptr )->CallRun( threadIndex );
        }
    }
}
```
and use it for both tasks

Reply:
fixed